src/common/survival_util.cc

```diff
@@ -0,0 +1,276 @@
+  grad = pdf * (1 - e_z) / (1 + e_z);
+  return grad;
+}
+
+double AFTLogistic::HessPDF(double z) {
+  double pdf;
+  double hess;
+  double w;
+  pdf = this->PDF(z);
+  w = std::exp(z);
+  hess = pdf * (w * w - 4 * w + 1) / ((1 + w) * (1 + w));
+  return hess;
+}
+
+double AFTExtreme::PDF(double z) {
+  double pdf;
+  double w;
+  w = std::exp(z);
+  pdf = w * std::exp(-w);
+  return pdf;
+}
+
+double AFTExtreme::CDF(double z) {
+  double cdf;
+  double w;
+  w = std::exp(z);
+  cdf = 1 - std::exp(-w);
+  return cdf;
+}
+
+double AFTExtreme::GradPDF(double z) {
+  double pdf;
+  double w;
+  double grad;
+
+  pdf = this->PDF(z);
+  w = std::exp(z);
+  grad = (1 - w) * pdf;
+  return grad;
+}
+
+double AFTExtreme::HessPDF(double z) {
+  double pdf;
+  double w;
+  double hess;
+  pdf = this->PDF(z);
+  w = std::exp(z);
+  hess = (w * w - 3 * w + 1) * pdf;
+  return hess;
+}
+
+
+double AFTLoss::Loss(double y_lower, double y_higher, double y_pred, double sigma) {
+  double pdf;
+  double cdf_u, cdf_l, z_u, z_l;
+  double cost;
+  if (y_lower == y_higher) {  // uncensored
+    z_l = (y_lower - y_pred) / sigma;
+    pdf = dist_->PDF(z_l);
+    cost = -std::log(pdf / (sigma * y_lower));
+  } else {  // censored; now check what type of censorship we have
+    if (std::isinf(y_higher)) {  // right-censored
+      cdf_u = 1;
+    } else {  // left-censored or interval-censored
+      z_u = (y_higher - y_pred) / sigma;
+      cdf_u = dist_->CDF(z_u);
+    }
+    if (std::isinf(y_lower)) {  // left-censored
+      cdf_l = 0;
+    } else {  // right-censored or interval-censored
+      z_l = (y_lower - y_pred) / sigma;
+      cdf_l = dist_->CDF(z_l);
+    }
+    cost = -std::log(cdf_u - cdf_l);
+  }
+  return cost;
+}
+
+double AFTLoss::Gradient(double y_lower, double y_higher, double y_pred, double sigma) {
+  double pdf_l;
+  double pdf_u;
+  double pdf;
+  double grad;
+  double z;
+  double z_u;
+  double z_l;
+  double cdf_u;
+  double cdf_l;
+  double gradient;
+  const double eps = 1e-12f;
+
+  if (y_lower == y_higher) {  // uncensored
+    z = (y_lower - y_pred) / sigma;
+    pdf = dist_->PDF(z);
+    grad = dist_->GradPDF(z);
+    gradient = grad / (sigma * pdf);
+  } else {  // censored; now check what type of censorship we have
+    if (std::isinf(y_higher)) {  // right-censored
+      pdf_u = 0;
+      cdf_u = 1;
+    } else {  // interval-censored or left-censored
+      z_u = (y_higher - y_pred) / sigma;
+      pdf_u = dist_->PDF(z_u);
+      cdf_u = dist_->CDF(z_u);
+    }
+    if (std::isinf(y_lower)) {  // left-censored
+      pdf_l = 0;
+      cdf_l = 0;
+    } else {  // interval-censored or right-censored
+      z_l = (y_lower - y_pred) / sigma;
+      pdf_l = dist_->PDF(z_l);
+      cdf_l = dist_->CDF(z_l);
+    }
+    gradient = (pdf_u - pdf_l) / (sigma * std::max(cdf_u - cdf_l, eps));
+  }
+
+  return gradient;
+}
+
+double AFTLoss::Hessian(double y_lower, double y_higher, double y_pred, double sigma) {
+  double z;
```

Comment:
Please add a summarization comment for these symbols, and link/names corresponding material of papers/books/notes ...

Reply:
Yes. I will add.